### PR TITLE
New Story Details Modal: Mandatory Field Functionality + UI Updates

### PR DESCRIPTION
**Pull request by BrittanyIRL**
## Context

Part of bigger feature for new story details modal, splitting up into chunks for faster review and QA where needed. 

## Summary

Wires up `useStory` to update story title and description, following same logic as existing excerpt and title updates in existing panels/inputs. 
Adds a 'required' label to the modal
More general structure. The modal still isn't connected to the editor yet. I'll do that next.  

## Relevant Technical Choices

- New label shared component for modal to distinguish between required and optional fields. Amy's good with the asterisk and tooltip. 
- Some constants.
- Trying to limit data getting passed around.
- New util for shared story updates - made one file for them for now, if this gets long or needs to be more specific i'll do that later. No use in over engineering from the start. We're going to have some overlapping functionality with these story updates and form components. There's _slight_ differences that made me want to just pull duplicative logic into utils to share (like the slug update) rather than try to force 2 different displays with extra logic in places (take the 'highlight' structure used in the existing panels). I think this keeps code separate where it makes sense and shared where the logic kicks in. I used the story title + slug on blur as an initial proof of concept for this patterning. 
- Initial unit tests set up - will do Karma set up with experiment set up next. 

## To-do

(See #10115 for full list) 

## User-facing changes

- Pulls the slug update functionality into a util. This impacts the story title on blur, which uses the same logic so it needs to be tested for regression. Testing instructions listed below are specifically for this. 

## Testing Instructions

**Slug Regression Testing**

The permalink in the document panel updates based on the story title IF there is no custom permalink set before updating the story title. So, with that in mind we need to just make sure that permalink updates are working as expected (it _should_, all I did was pull it out into a util and update the currentSlug equality check to turn it into a number to check properly against the current story id as a fallback). 
1. If a story is published without a story title or custom setting a permalink the permalink will be the story id 
2. If a story title is given, when user clicks or tabs out of the story title input the permalink is updated with a slug version of the title ('hold on to your butts!' turns into 'hold-on-to-your-butts') 
3. If the permalink is updated on it's own ('jurassic-park-rules') and then the story title is updated, the permalink remains as it was 
4. If a permalink is removed and then the story title is updated or the story is published, the permalink is a slug based on the title or if no title (shame on you for not titling your story) it'll be the story id. 

Otherwise, you can run storybook locally. Now the story details modal looks like this: 

Screen Shot 2022-01-28 at 2 56 47 PM

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Partially addresses #10115

### CODE REVIEW

File: packages/story-editor/src/utils/storyUpdates.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ * @param {Function} args.updateStory The callback to useStory's action for updating the story with new slug.
+ * @return {void}
+ */
+export const updateSlug = ({
+  currentSlug,
+  currentTitle,
+  storyId,
+  updateStory,
+}) => {
+  if (!currentSlug || parseInt(currentSlug) === storyId) {
```
Comment (BrittanyIRL) [inline packages/story-editor/src/utils/storyUpdates.js]: This line is different from the original check to update the slug that was in the `title` onBlur that I pulled out into a util. When I was testing it I noticed that `currentSlug === storyId` is never going to be true without `parseInt` because the slug's a string and storyId is a number. This is why I want to specifically do regression testing on the slug updates. 

@miina is there any history here with this check that I'm missing or context that makes this a bad update?

Comment (miina) [reply packages/story-editor/src/utils/storyUpdates.js]: Perhaps the check without `parseInt` was for when the story was still new and hadn't been saved at all, and then changing the slug, could it be? Or perhaps some non-standard permalink setting?
I'll double check tomorrow if that's not the case (finished the day already).

Comment (BrittanyIRL) [reply packages/story-editor/src/utils/storyUpdates.js]: If the story is new and hasn't been saved at all then the slug's empty. 

The story id is used as a fallback if there's no story title. So, for instance, if I focus the story title and then tab out of it, there's still no viable story title in which to make a slug out of so the story id is used as a fallback. 

Where the parseInt comes into play is if the current slug is equal to the story id on a story title update then we should check and see if there's an update, otherwise the slug isn't updated every time the story title changes (which is consistent with staging). I honestly think this shoulda been a parseInt all along but it's such an edge case it wasn't really noticed. I did set up unit tests now so we can be confident in when this is updating and how.

**Review by BrookeGraham — APPROVED** [review]
Curious if Mina has any more insight on the slug. It all sounds (and looks) good to me. Tests look great too!

File: packages/story-editor/src/utils/storyUpdates.js (hunk 1 above)
Comment (miina) [reply packages/story-editor/src/utils/storyUpdates.js]: Ah, okay, got it. The only place where I can see this could now cause issues is if the slug starts with the story ID.
E.g. the story ID is 3 and the slug is `3-things-you-can-do`, then blurring the title would overwrite that slug, if I understand correctly.

Comment (swissspidy) [reply packages/story-editor/src/utils/storyUpdates.js]: `Number()` seems preferable in that case over `parseInt`

File: packages/design-system/src/components/typography/headline/index.js [hunk 2]
```diff
@@ -46,7 +46,7 @@ export const Headline = styled.h1`
 `;
 
 Headline.propTypes = {
-  as: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'a']),
+  as: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'a', 'label']),
```
Comment (samwhale) [inline packages/design-system/src/components/typography/headline/index.js:49]: 💭  Doesn't the `label` already exist as an option of `<Text>`?

https://github.com/GoogleForCreators/web-stories-wp/blob/d16f28fb45d808bf5abef2a5ae2c69feb5a63f66/packages/design-system/src/components/typography/text/index.js#L88-L94

Does the headline need this option?

File: packages/story-editor/src/components/publishModal/header/index.js [hunk 3]
```diff
@@ -48,31 +49,41 @@ const PublishButton = styled(Button)`
...
-      </PublishButton>
+      </Button>
       <Headline
         as="h2"
         size={THEME_CONSTANTS.TYPOGRAPHY.PRESET_SIZES.XXX_SMALL}
       >
         {__('Story Details', 'web-stories')}
       </Headline>
-      <Button
+      <PublishButton
```
Comment (samwhale) [inline packages/story-editor/src/components/publishModal/header/index.js:70]: 👍

File: packages/story-editor/src/components/publishModal/mainContent/formLabel.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    </Tooltip>
+  ) : (
+    <Headline
+      as="label"
+      size={THEME_CONSTANTS.TYPOGRAPHY.PRESET_SIZES.XX_SMALL}
+      htmlFor={htmlFor}
+    >
+      {copy}
+    </Headline>
+  );
```
Comment (samwhale) [inline packages/story-editor/src/components/publishModal/mainContent/formLabel.js:54]: 💭  what are your thoughts on adding this directly to the `TextArea` in the design system? The `TextArea` component has an option to render a label that matches the designs but no option to render the tooltip on the label

Wonder if that would be good to add 🤔

File: packages/story-editor/src/components/publishModal/test/mainContent.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+  const mockInputValues = {
+    [INPUT_KEYS.EXCERPT]:
+      'A voyage to the great beyond, where none have gone and going back is impossible.',
+    [INPUT_KEYS.TITLE]: "David Bowman's Odyssey",
+    [INPUT_KEYS.SLUG]: '2001-space-odyssey',
+  };
+  afterEach(() => {
+    mockHandleUpdateStoryInfo.mockReset();
+    mockHandleUpdateSlug.mockReset();
```
Comment (samwhale) [inline packages/story-editor/src/components/publishModal/test/mainContent.js]: nit: could write `jest.resetAllMocks()` here if you want. Would also be easier if more mocks are added in the future

Comment (samwhale) [reply packages/story-editor/src/components/publishModal/test/mainContent.js]: Also - doesn't seem like these mocks are returning anything here. We could use `clearAllMocks` instead: 

https://jestjs.io/docs/mock-function-api#mockfnmockreset
image

File: packages/story-editor/src/utils/test/storyUpdates.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+ */
+import { updateSlug } from '../storyUpdates';
+
+describe('storyUpdates', () => {
+  describe('updateSlug', () => {
+    const updateStory = jest.fn();
+
+    afterEach(() => {
+      updateStory.mockReset();
+    });
```
Comment (samwhale) [inline packages/story-editor/src/utils/test/storyUpdates.js:27]: same here!

**Review by samwhale — APPROVED** [review]
Great tests, very straight forward. Modal is looking clean too.

File: packages/story-editor/src/components/publishModal/mainContent/formLabel.js (hunk 4 above)
Comment (BrittanyIRL) [reply packages/story-editor/src/components/publishModal/mainContent/formLabel.js:54]: Since this is a 1 off pattern (as far as I'm aware), i think keeping it out of the design system for now is better.

File: packages/design-system/src/components/typography/headline/index.js (hunk 2 above)
Comment (BrittanyIRL) [reply packages/design-system/src/components/typography/headline/index.js:49]: 😢 design has the line spacing matching the `Headline` instead of the `Text` so i added a label option there.  I should check with Amy though, maybe it was an accident or not important.

**Comment by felipebochehin87** [comment]
Tested against https://stories-qa-wordpress-amp.pantheonsite.io/wp-admin/, using PR #10403.

1. Published story with both title and permalink empty - story id is set as permalink
2. Added a title and pressed tab - permalink is set to the same as the title
3. Changed permalink and changed title - permalink remains as it was
4. Removed permalink and updated story title - permalink is the story title
5. Removed permalink and story title - permalink is story id


https://user-images.githubusercontent.com/96055551/152569606-9b6a2fd3-1331-46bc-9524-4ad3d9873f52.mp4
